bkbit/scripts/gfftranslator.py

```diff
@@ -4,133 +4,346 @@
+    genomeAnnot (GenomeAnnotation): reference to GenomeAnnotation object 
+    orgTaxon (OrganismTaxon): reference to OrganismTaxon object 
+    taxon_id (str): taxon id 
 
-    # ATTRIBUTES FOR GENE ANNOTATION OBJECT:
-    cols = ['genome_annotation_label','gene_identifier_prefix','gene_local_unique_identifier','name','description', 'gene_biotype', 'taxon_unique_identifier']
+    Returns:
+    gene_annotations (List[GeneAnnotations]): list of gene annotations created
 
-    # create output dataframe
-    out_df = df_gene[out_cols]
-    out_df.columns = cols
+    """
+    gene_annotations = []
 
-    fname = 'gene_annotation_%s-%s-%s.csv' % (authority, str(taxon_local_unique_id), str(version))
-    file = os.path.join(output_dir, fname)
-    out_df.to_csv(file, index=False)
+    for row in df.itertuples():
+        attributes = dict([x.split('=') for x in row.attributes.split(';')])
+        description = attributes.get("description")
+        if description is not None:
+            description = urllib.parse.unquote(description)
+            description = re.sub(r"\[source .*?\]", "", description)
+        stable_id = attributes['gene_id'].split('.')[0]
+        molecular_type = kbmodel.BioType.protein_coding if attributes.get('biotype') == kbmodel.BioType.protein_coding.value else kbmodel.BioType.noncoding
+
+
+        curr_gene_annot = kbmodel.GeneAnnotation(id = ENSEMBL_GENE_ID_PREFIX + ':' + stable_id , symbol = attributes.get('Name'), name = attributes.get('Name'), referenced_in = genomeAnnot.id, 
+                                                molecular_type = molecular_type, source_id = stable_id, in_taxon = [orgTaxon.id], 
+                                                in_taxon_label = TAXON_SCIENTIFIC_NAME[taxon_id], description = description, synonym = None)
+        gene_annotations.append(curr_gene_annot)
+    return remove_ensembl_duplicates(gene_annotations)
+
+
+def remove_ensembl_duplicates(gene_annotations):
+    """
+    Removes duplicates in a list of gene annotations based on [id, name, description, molecular_type]. 
+
+    Parameters:
+    gene_annotations (list): A list of gene annotations.
 
+    Returns:
+    list[GeneAnnotations]: A list of unique gene annotations.
 
-def create_gene_annotation_objects(file, genomeAnnot, orgTaxon):
-    """Initializes GeneAnnotation objects using data from .csv file.
+    """
+    #! LONG TERM TODO: make GeneAnnotation hashable so we can use set(). Add def __hash__(self) to GeneAnnotation class in kbmodel.py
+    unique_annotations = dict()
+    seen_annotations = set()
+
+    for annotation in gene_annotations:
+        annotation_key = (annotation.id, annotation.name, annotation.description, annotation.molecular_type)
+
+        if annotation_key not in seen_annotations:
+            unique_annotations[annotation.id] = annotation
+            seen_annotations.add(annotation_key)
+        else:
+            try:
+                unique_annotations.pop(annotation.id)
```

Comment:
The document is the for ensembl we want to remove drop all duplicate rows.

Let's take an example: if the gff3 contained 5 rows with genes: gene:A, gene:B, gene:C, gene:A, gene:A. The resulting unique set of genes should only include gene:B and gene:C.

Given the example (and if my assumptions of how we handle duplicates ensembl genes is correct) if we walk through the current code; the first occurrence of gene:A will be added to `unique_annotations` since it is not in `seen_annotations` yet; therefore, when you iterate over the second occurrence of gene:A you also have to remove it from `unique_annotations`.

Reply:
Okay. That is not how I interpreted "drop all duplicate rows" in the requirements document. I thought it meant "keep only the first instances of any set of duplicate rows", so you would have an entry for gene:A in your example.

We should probably clarify this with Lydia (unless you already have).

Reply:
@puja-trivedi, @danielsf 

Sorry about the confusion. When I wrote "drop all duplicate rows", I was thinking the default behavior of pandas drop duplicate - which keep the first instance and drop the other instances.

I have also updated the WORD doc with this clarification